File: server/Sources/RevaServer/SnapshotMetadata.swift
// Purpose: Validate additive date/title provenance without reinterpreting legacy clinical snapshots.
// Inputs: Structurally validated snapshot objects with optional record and recording metadata.
// Outputs: Acceptance or a sanitized validation error; original timestamps remain unchanged.
// Side effects: None. Missing/null metadata remains compatible with older clients.

import Foundation
import Vapor

// MARK: - Optional provenance fields only; unrelated legacy snapshot fields remain untouched
extension Validation {
    static func snapshotMetadata(_ snapshot: [String: JSONValue]) throws {
        func optionalEnum(_ value: JSONValue?, choices: [String]) -> Bool {
            guard let value, value != .null else { return true }
            guard case .string(let string) = value else { return false }
            return choices.contains(string)
        }
        func optionalDate(_ value: JSONValue?) -> Bool {
            guard let value, value != .null else { return true }
            guard case .string(let string) = value else { return false }
            return validMetadataDate(string)
        }
        for key in ["records", "recordings"] {
            guard case .array(let values) = snapshot[key] else { continue }
            for value in values {
                guard case .object(let object) = value else { continue }
                let valid: Bool
                if key == "records" {
                    valid =
                        optionalEnum(
                            object["dateSource"], choices: ["document", "observed", "recorded", "added"])
                        && optionalDate(object["summaryGeneratedAt"])
                } else {
                    valid =
                        optionalEnum(object["titleSource"], choices: ["user", "date", "ai"])
                        && ["capturedAt", "savedAt", "aiSummaryGeneratedAt"].allSatisfy {
                            optionalDate(object[$0])
                        }
                }
                guard valid else {
                    throw Abort(
                        .badRequest,
                        reason: "Invalid optional record or recording date/title provenance metadata.")
                }
            }
        }
    }

    // MARK: - Strict ISO date/instant syntax and Gregorian validity prevent implicit rollover
    static func validMetadataDate(_ value: String) -> Bool {
        guard value.utf8.count <= 40,
            value.range(
                of:
                    #"^\d{4}-\d{2}-\d{2}(?:T(?:[01]\d|2[0-3]):[0-5]\d:[0-5]\d(?:\.\d{1,9})?(?:Z|[+-](?:[01]\d|2[0-3]):[0-5]\d))?\z"#,
                options: .regularExpression) != nil
        else { return false }
        let parts = value.prefix(10).split(separator: "-")
        guard parts.count == 3, let year = Int(parts[0]), let month = Int(parts[1]), let day = Int(parts[2]),
            year >= 1, (1...12).contains(month)
        else { return false }
        let leap = year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
        let days = [31, leap ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
        return (1...days[month - 1]).contains(day)
    }
}

File: backend/snapshot-metadata.mjs
// Purpose: Validate only optional date/title provenance added to synchronized records and recordings.
// Inputs: Structurally validated snapshot arrays containing untrusted optional metadata.
// Outputs: A boolean acceptance decision without rewriting, inferring or adding timestamps.
// Side effects: None. Missing/null fields and unrelated legacy snapshot content remain accepted.

// MARK: - Bounded ISO calendar dates and instants, including explicit numeric offsets
export function validMetadataDate(value) {
  if (typeof value !== "string" || value.length > 40) return false;
  const match =
    /^(\d{4})-(\d{2})-(\d{2})(?:T(?:[01]\d|2[0-3]):[0-5]\d:[0-5]\d(?:\.\d{1,9})?(?:Z|[+-](?:[01]\d|2[0-3]):[0-5]\d))?$/.exec(
      value,
    );
  if (!match || match[0] !== value) return false;
  const [, yearText, monthText, dayText] = match;
  const year = Number(yearText),
    month = Number(monthText),
    day = Number(dayText);
  if (year < 1 || month < 1 || month > 12) return false;
  const leap = year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
  const days = [31, leap ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  return (
    day >= 1 && day <= days[month - 1] && Number.isFinite(Date.parse(value))
  );
}

// MARK: - Optional provenance metadata is typed without tightening the rest of a snapshot
export function validSnapshotMetadata(snapshot) {
  const optionalEnum = (value, choices) =>
    value == null || choices.includes(value);
  const optionalDate = (value) => value == null || validMetadataDate(value);
  return (
    snapshot.records.every(
      (record) =>
        optionalEnum(record.dateSource, [
          "document",
          "observed",
          "recorded",
          "added",
        ]) && optionalDate(record.summaryGeneratedAt),
    ) &&
    snapshot.recordings.every(
      (recording) =>
        optionalEnum(recording.titleSource, ["user", "date", "ai"]) &&
        ["capturedAt", "savedAt", "aiSummaryGeneratedAt"].every((field) =>
          optionalDate(recording[field]),
        ),
    )
  );
}
